task-1/src/Shape.ts

```diff
@@ -0,0 +1,37 @@
+import { Point } from './Point';
+
+export abstract class Shape {
+  protected color: string;
+  protected filled: boolean;
+  protected points: Point[];
+
+  constructor(points: Point[], color?: string, filled?: boolean);
+  constructor(points: Point[], color: string, filled: boolean) {
+    if (points.length < 3) {
+      throw new Error;
+    }
+
+    this.points = points;
+    this.color = color ?? "green";
+    this.filled = filled ?? true;
+  }
+
+  private isLastIndex = (index: number, arrayLength: number) => index === arrayLength - 1;
+
+  private getTotalPoints(): string {
+    return this.points.map((point: Point) => point.toString()).join(', ');
+  }
+
+  toString(): string {
+    return `A Shape with color of ${this.color} and ${this.filled ? 'filled' : 'not filled'}. Points: ${this.getTotalPoints()}.`
+  }
+
+  getPerimeter(): number {
```

Comment:
This code is actually duplicated in two places: here, and inside of Triangle.getType.
To eliminate this, we can add one more method - pointsToDistance (or whatever it calls)
```ts
  protected pointsToDistance(): Array<number> {
    return this.points.map((pointA, pointAIndex) => (
        pointA.distance(this.getNextPoint(pointAIndex))
    ))
  }
```
And, we can modify getPerimeter to just
```ts
  getPerimeter(): number {
    return this.pointsToDistance().reduce((perimeter, distance) => perimeter + distance)
  }
```

Also, a little mark. It's not necessary inside of reduce to write `accumulator += number`. You can just write `accumulator + number`, the assignment is redundant